# Homework Introduction to Python

$\textbf{Ubah nama di bawah yaa!}$

- Nama: Deni Yuniawan
- Batch: 45

In [9]:
import numpy as np

# please import as much as you need

# Normal

## 1. Operasi Variabel

> Misal kita memiliki toko sepatu. Modal membeli 1 kodi sepatu adalah Rp 1.000.000. Setiap sepatu dijual seharga Rp 80.000. Buatlah variabel modal, kuantitas, dan harga_jual sesuai informasi di atas. Hitung profit yang didapatkan jika semua sepatu berhasil terjual, dan masukkan ke variabel profit.

Saran langkah-langkah pengerjaan
1. Convert terlebih dahulu kodi dalam jumlah barang (1 kodi = xyz buah sepatu)
2. Gunakan operasi matematika sederhana untuk menghitung profit menggunakan tiga buah variabel yang sudah ditentukan

In [10]:
# Modal membeli 1 kodi sepatu (dalam Rupiah)
modal_per_kodi = 1000000

# Harga jual per sepatu (dalam Rupiah)
harga_jual_per_sepatu = 80000

# Jumlah sepatu dalam 1 kodi
jumlah_sepatu_per_kodi = 20

# Kuantitas kodi sepatu yang dibeli
kuantitas_kodi = 1

# Convert kodi menjadi jumlah sepatu
kuantitas = kuantitas_kodi * jumlah_sepatu_per_kodi

# Hitung modal total
modal_total = modal_per_kodi * kuantitas_kodi

# Hitung total pendapatan dari penjualan semua sepatu
total_pendapatan = kuantitas * harga_jual_per_sepatu

# Hitung profit menggunakan operasi matematika sederhana
profit = total_pendapatan - modal_total

# Cetak hasil
print("Modal Total:", modal_total, "Rupiah")
print("Kuantitas Sepatu:", kuantitas, "buah")
print("Total Pendapatan:", total_pendapatan, "Rupiah")
print("Profit:", profit, "Rupiah")

Modal Total: 1000000 Rupiah
Kuantitas Sepatu: 20 buah
Total Pendapatan: 1600000 Rupiah
Profit: 600000 Rupiah


Jadi, dari profit di atas, dapat disimpulkan bahwa...

* Modal Total: 1000000 Rupiah 
* Kuantitas Sepatu: 20 buah 
* Total Pendapatan: 1600000 Rupiah 
* Profit: 600000 Rupiah

## 2. Package Python dan Dictionary

> Kita akan coba mengakses data scientific article dari PLOS corpus
    Menggunakan API request dengan kode berikut :
```
import requests
response = requests.get('http://api.plos.org/search?q=title:"Drosophila" AND body:"RNA"&fl=id,abstract')
```
**perhatikan tanda kutipnya tidak salah ketika meng-copy ke dalam python

Pertanyaannya :
Convert response.content dari request tersebut dari JSON Format ke dalam Dictionary, dan print-kan key di dalamnya apa saja

Saran langkah-langkah pengerjaan
1. Ikuti code dari soal, kemudian ekstrak value dari variable response menggunakan dot (.) content menjadi sebuah variabel baru (dictionary)
2. Gunakan library json, convert hasil content menggunakan json.loads
3. Print keys dari dictionary tersebut


In [11]:
# your code goes hemport requests
import requests
import json

# Mengirim permintaan API
response = requests.get('http://api.plos.org/search?q=title:"Drosophila" AND body:"RNA"&fl=id,abstract')
dataset = response.content

#parsingjson
dataset_parse = json.loads(dataset)

#printkey di dalam dict dataset parse
print(dataset_parse.keys())
#printnestedkey
print(dataset_parse['response'].keys())
#printmorenestedkey
print(dataset_parse['response']['docs'][0].keys())

dict_keys(['response'])
dict_keys(['numFound', 'start', 'docs'])
dict_keys(['id', 'abstract'])


Jadi, dari url link di atas, kami menyimpulkan bahwa memiliki keys ...
- '*Response*'

# Intermediate

## 3. Operasi dalam List

> Buatlah list dengan 10 nama kota di indonesia, kemudian:
1. Urutkan nama kota di dalam list berdasarkan abjad (a)
2. Ambil 4 kota pertama dari list tersebut (yang telah diurutkan) (b)
3. Ambil 3 kota terakhir dari list tersebut (yang telah diurutkan) (c)

Saran langkah-langkah pengerjaan
1. Tidak ada ketentuan nama kota (misal berdasarkan provinsi X atau pulau Y), dibebaskan untuk memilih. Buat nama-nama kota tersebut dalam sebuah list.
2. Gunakan fungsi sorted(X) untuk mengurutkan secara kecil-ke-besar (sesuai abjad), dimana X adalah contoh variabel list pada langkah (a)
3. Gunakan hasil poin (a), urutkan nama kota di dalam list berdasarkan abjad, untuk mengerjakan poin (b) dan (c).

In [12]:
# Membuat list dengan 10 nama kota di Indonesia
nama_kota = ['Jakarta', 'Surabaya', 'Balikpapan', 'Medan', 'Samarinda', 
             'Makassar', 'Denpasar', 'Tangerang', 'Depok', 'Malang']

# Mengurutkan nama kota berdasarkan abjad
nama_kota_urutan = sorted(nama_kota)

# Mengambil 4 kota pertama dari list yang telah diurutkan
kota_awal = nama_kota_urutan[:4]

# Mengambil 3 kota terakhir dari list yang telah diurutkan
kota_akhir = nama_kota_urutan[-3:]

# Menampilkan hasil
print("a. Nama kota diurutkan berdasarkan abjad:", nama_kota_urutan)
print("b. 4 Kota pertama:", kota_awal)
print("c. 3 Kota terakhir:", kota_akhir)

a. Nama kota diurutkan berdasarkan abjad: ['Balikpapan', 'Denpasar', 'Depok', 'Jakarta', 'Makassar', 'Malang', 'Medan', 'Samarinda', 'Surabaya', 'Tangerang']
b. 4 Kota pertama: ['Balikpapan', 'Denpasar', 'Depok', 'Jakarta']
c. 3 Kota terakhir: ['Samarinda', 'Surabaya', 'Tangerang']


Jadi, dari operasi list di atas, dapat disimpulkan bahwa...

a. Nama kota diurutkan berdasarkan abjad: ['Balikpapan', 'Denpasar', 'Depok', 'Jakarta', 'Makassar', 'Malang', 'Medan', 'Samarinda', 'Surabaya', 'Tangerang']

b. 4 Kota pertama: ['Balikpapan', 'Denpasar', 'Depok', 'Jakarta']

c. 3 Kota terakhir: ['Samarinda', 'Surabaya', 'Tangerang']

## 4. Function dalam List

> Team sales memiliki list total transaksi dari 10 merchant pada bulan Juni 2020, yaitu:
```
total_transaksi = [ 19, 35, 23, 25, 10, 15, 33, 45, 21, 30 ]
```
1. Trx minimum
2. Rata-rata
3. Trx maksimal

Saran langkah-langkah pengerjaan
1. Copy variabel total_transaksi yang ada di soal ke notebook kita (atau ketik ulang boleh)
2. Kemudian, hitung menggunakan function operasi dalam list yang telah dipelajari di kelas pada topik list. Print hasilnya setiap poin (a), (b), dan (c)

In [13]:
# Copy variabel total_transaksi
total_transaksi = [19, 35, 23, 25, 10, 15, 33, 45, 21, 30]

# Trx minimum
trx_minimum = min(total_transaksi)
print("a. Transaksi minimum:", trx_minimum)

# Rata-rata
rata_rata = sum(total_transaksi) / len(total_transaksi)
print("b. Rata-rata transaksi:", rata_rata)

# Trx maksimal
trx_maksimal = max(total_transaksi)
print("c. Transaksi maksimal:", trx_maksimal)

a. Transaksi minimum: 10
b. Rata-rata transaksi: 25.6
c. Transaksi maksimal: 45


Jadi, dari hasil operasi list di atas, dapat disimpulkan bahwa nilai minimal, max, rata-rata...

- a. Transaksi minimum: 10 
- b. Rata-rata transaksi: 25.6 
- c. Transaksi maksimal: 45

# Advanced

## 5. NumPy Array

> Buatlah array dari 5 bilangan ganjil pertama (dimulai dari 1), Kemudian lakukan perintah berikut :
1. Kali-kan array tersebut dengan 10
2. Bagi-kan array (hasil dari bagian a) dengan 3
3. Kali-kan array (hasil dari bagian b) dengan array baru :
`[-45, 43, 3, 56, 5]`
4. Carilah pada index ke-berapa nilai tertinggi dan terendah dari array (hasil dari bagian c)

Saran langkah-langkah pengerjaan
1. HATI-HATI dalam pengerjaan, karena apabila poin sebelumnya salah, maka selanjutnya akan kebawa salah (misal (b) salah, maka poin (c) sampai (d) akan salah)
2. Untuk array 5 bilangan ganjil pertama boleh manual boleh menggunakan fungsi, dimulai dari 1
3. Ikuti perintah soal, gunakan numpy array, jangan list. Sesuaikan dengan permintaan soal pada setiap butir.



In [14]:
import numpy as np

# Buat array dari 5 bilangan ganjil pertama
array_ganjil = np.array([1, 3, 5, 7, 9])

# a. Kali-kan array tersebut dengan 10
array_kali_10 = array_ganjil * 10
print("a. Hasil perkalian array dengan 10:", array_kali_10)

a. Hasil perkalian array dengan 10: [10 30 50 70 90]


In [15]:
# b. Bagi-kan array (hasil dari bagian a) dengan 3
array_bagi_3 = array_kali_10 / 3
print("b. Hasil pembagian array dengan 3:", array_bagi_3)

b. Hasil pembagian array dengan 3: [ 3.33333333 10.         16.66666667 23.33333333 30.        ]


In [16]:
# c. Kali-kan array (hasil dari bagian b) dengan array baru
array_baru = np.array([-45, 43, 3, 56, 5])
array_kali = array_bagi_3 * array_baru
print("c. Hasil perkalian array dengan array baru:", array_kali)

c. Hasil perkalian array dengan array baru: [-150.          430.           50.         1306.66666667  150.        ]


In [17]:
# d. Carilah pada index ke-berapa nilai tertinggi dan terendah dari array (hasil dari bagian c)
index_max = np.argmax(array_kali)
index_min = np.argmin(array_kali)

print("d. Index nilai tertinggi:", index_max)
print("   Index nilai terendah:", index_min)
print("   Nilai tertinggi:", array_kali[index_max])
print("   Nilai terendah:", array_kali[index_min])

d. Index nilai tertinggi: 3
   Index nilai terendah: 0
   Nilai tertinggi: 1306.6666666666665
   Nilai terendah: -150.0


Jadi, dari hasil operasi numpy di atas, dapat disimpulkan bahwa..

- a. Hasil perkalian array dengan 10: [10 30 50 70 90] 
- b. Hasil pembagian array dengan 3: [ 3.33333333 10. 16.66666667 23.33333333 30. ] 
- c. Hasil perkalian array dengan array baru: [-150. 430. 50. 1306.66666667 150. ] 
- d. Index nilai tertinggi: 3 Nilai tertinggi: 1306.6666666666665 

- Index nilai terendah: 0 
- Nilai terendah: -150.0

# Terimakasih, salam 